modules/sdk-coin-sui/src/lib/transaction.ts

```diff
@@ -1,16 +1,360 @@
+  Single: 'Transaction',
+  Batch: 'vector<Transaction>',
+});
+
+bcs.registerStructType('TransactionData', {
+  kind: 'TransactionKind',
+  sender: 'SuiAddress',
+  gasPayment: 'SuiObjectRef',
+  gasPrice: 'u64',
+  gasBudget: 'u64',
+});
+
+bcs.registerType(
+  'SignatureData',
+  (writer, str) => {
+    const bytes = Array.from(decodeStr(str, 'base64'));
+    return writer.writeVec(bytes, (writer, el) => writer.write8(el));
+  },
+  (reader) => {
+    const bytes = reader.readVec((reader) => reader.read8());
+    return encodeStr(new Uint8Array(bytes), 'base64');
+  }
+);
+
+bcs.registerEnumType('Signature', {
+  Ed25519SuiSignature: 'SignatureData',
+  Secp256k1SuiSignature: 'SignatureData',
+});
+
+bcs.registerStructType('SenderSignedData', {
+  data: 'TransactionData',
+  txSignature: 'vector<u8>',
+});
+
+const UNAVAILABLE_TEXT = 'UNAVAILABLE';
+export const TYPE_TAG = Array.from('TransactionData::').map((e) => e.charCodeAt(0));
+
+export type SuiObjectRef = {
+  /** Hex code as string representing the object id */
+  objectId: string;
+  /** Object version */
+  version: number;
+  /** Base64 string representing the object digest */
+  digest: string;
+};
+
+export type SuiTransaction = {
+  sender: string;
+  kind: PayTransactionKind;
+  gasBudget: number;
+  gasPrice: number;
+  gasPayment?: SuiObjectRef;
+};
 
 export class Transaction extends BaseTransaction {
+  private _suiTransaction: SuiTransaction;
+  private _signature;
+  private _publicKey;
+
+  constructor(_coinConfig: Readonly<CoinConfig>) {
+    super(_coinConfig);
+  }
+
+  get suiTransaction(): SuiTransaction {
+    return this._suiTransaction;
+  }
+
+  setSuiTransaction(value: SuiTransaction): void {
+    this._suiTransaction = value;
+  }
+
+  /** @inheritDoc **/
+  get id(): string {
+    // Solana transaction ID === first signature: https://docs.solana.com/terminology#transaction-id
+    if (this._signature) {
+      return base58.encode(this._signature);
+    } else {
+      return UNAVAILABLE_TEXT;
+    }
+  }
+
+  getInputCoins(): SuiObjectRef[] {
+    return this.suiTransaction.kind.Single.Pay.coins;
+  }
+
+  /** @inheritdoc */
   canSign(key: BaseKey): boolean {
-    return false;
+    return true;
   }
 
+  /** @inheritdoc */
   toBroadcastFormat(): string {
-    throw new Error('Method not implemented.');
+    if (!this._suiTransaction) {
+      throw new InvalidTransactionError('Empty transaction');
+    }
+    return this.serialize();
   }
 
+  /** @inheritdoc */
   toJson(): TxData {
-    throw new Error('Method not implemented.');
+    if (!this._suiTransaction) {
+      throw new ParseTransactionError('Empty transaction');
+    }
+
+    return {
+      id: this._id,
+      gasBudget: this._suiTransaction.gasBudget,
+      gasPrice: this._suiTransaction.gasPrice,
+      kind: this._suiTransaction.kind,
+      sender: this._suiTransaction.sender,
+      gasPayment: this._suiTransaction.gasPayment,
```

Comment:
can we include things like `recipients`, `amounts`, and `inputCoins`?

Reply:
`recipients`, `amounts` and `coins` are already in `kind.Single.Pay`, if add it here, would it be redundant?

Reply:
kind.Single.Pay is sui's own structure, maybe we can flatten it to easy understand.

Reply:
Simplified to just `PayTx` with fields `recipients`, `amounts` and `coins`